File: week-11/protoseo/호텔 방 배정.java
import java.util.*;

class Solution {
    Map<Long, Long> parents = new HashMap<>();
    public long[] solution(long k, long[] room_number) {
        int n = room_number.length;
        long[] answer = new long[n];

        for (int i = 0; i < n; i++) {
            long emptyRoom = find(room_number[i]);
            answer[i] = emptyRoom;
            union(emptyRoom, emptyRoom + 1);
        }
        return answer;
    }

    public long find(long a) {
        if (!parents.containsKey(a)) {
            return a;
        }
        long parent = find(parents.get(a));
        parents.put(a, parent);
        return parent;
    }

    public void union(long a, long b) {
        long pa = find(a);
        long pb = find(b);
        parents.put(pa, pb);
    }
}

File: week-11/pushedrumex/호텔방 배정.py
import sys
from collections import defaultdict

sys.setrecursionlimit(10000)

def find(parent, node):
    if parent[node] == 0:
        parent[node] = node + 1
        return node
    parent[node] = find(parent, parent[node])
    return parent[node]
        
def solution(k, room_number):
    parent = defaultdict(int)
    answer = []
    for n in room_number:
        answer.append(find(parent, n))
    return answer